lib/trino-filesystem/src/main/java/io/trino/filesystem/Location.java

```diff
@@ -0,0 +1,209 @@
+     * This should only be used once, as recursive calls for blob paths may lead to incorrect results.
+     *
+     * @throws IllegalStateException if the location is not a valid file location
+     */
+    public Location parentDirectory()
+    {
+        // todo should this only be allowed for file locations?
+        verifyValidFileLocation();
+        checkState(!path.isEmpty(), "root location does not have parent: %s", location);
+
+        int lastIndexOfSlash = path.lastIndexOf('/');
+        if (lastIndexOfSlash < 0) {
+            String newLocation = location.substring(0, location.length() - path.length() - 1);
+            return new Location(newLocation, scheme, userInfo, host, "");
+        }
+
+        String newPath = path.substring(0, lastIndexOfSlash);
+        String newLocation = location.substring(0, location.length() - (path.length() - newPath.length()));
+        return new Location(newLocation, scheme, userInfo, host, newPath);
+    }
+
+    /**
+     * Creates a new location by appending the given path element to the current path.
+     * A slash will be added between the current path and the new path element if needed.
+     *
+     * @throws IllegalArgumentException if the new path element is empty or starts with a slash
+     */
+    public Location appendPath(String newPathElement)
+    {
+        checkArgument(!newPathElement.isEmpty(), "newPathElement is empty");
+        checkArgument(!newPathElement.startsWith("/"), "newPathElement starts with a slash: %s", newPathElement);
+
+        if (path.isEmpty()) {
+            // empty path may or may not have a location that ends with a slash
+            return new Location(
+                    location + (location.endsWith("/") ? "" : "/") + newPathElement,
+                    scheme,
+                    userInfo,
+                    host,
+                    newPathElement);
+        }
+
+        if (!path.endsWith("/")) {
+            newPathElement = "/" + newPathElement;
+        }
+        return new Location(location + newPathElement, scheme, userInfo, host, path + newPathElement);
+    }
+
+    /**
+     * Verifies the location is valid for a file reference.  Specifically, the path must not be empty, must not end with a slash or whitespace.
+     *
+     * @throws IllegalStateException if the location is not a valid file location
+     */
+    public void verifyValidFileLocation()
+    {
+        // TODO: should this be IOException?
+        // file path must not be empty
+        checkState(!path.isEmpty() && !path.equals("/"), "File location must contain a path: %s", location);
+        // file path cannot end with a slash
+        checkState(!path.endsWith("/"), "File location cannot end with '/': %s", location);
+        // file path cannot end with whitespace
+        checkState(!isWhitespace(path.charAt(path.length() - 1)), "File location cannot end with whitespace: %s", location);
```

Comment:
Maybe we should put quotes around the `%s` here, since there is whitespace

Reply:
Should we disallow leading or trailing whitespace in the parsing code? I can't see how trailing whitespace is valid for a path.

Reply:
Maybe.  The original code tried to put a few constraints as possible on paths.  We can change this in follow up work.